src/autograd/Functions.cpp

```diff
@@ -117,6 +117,35 @@ namespace af {
             return Variable(result, false);
         }
 
+        Variable select_index(const Variable &input, const Variable &idx)
+        {
+            af::array result = input.array()(idx.array());
+            af::array mask   = af::constant(0, input.dims());
+            mask(idx.array()) = 1;
+
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+                auto grad = inputs[2].array();
+                auto grad_mask = af::where(grad);
+                grad(grad_mask) *= grad_output.array();
+
+                inputs[0].addGrad(Variable(grad, false));
+            };
+            return Variable(result, {input, idx, Variable(mask, false)}, grad_func);
+        }
+
+        Variable set_index(const Variable &input, const Variable &idx, const Variable &vals)
+        {
+            af::array result = input.array();
+            result(idx.array()) = vals.array();
+            af::array mask   = af::constant(1, input.dims(), s32);
+            mask(idx.array()) = 0;
+
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+                inputs[0].addGrad(inputs[3] * grad_output);
+            };
+            return Variable(result, {input, idx, vals, Variable(mask, false)}, grad_func);
```

Comment:
It is better to use just index and do the masking inside the grad_func. No reason to do the masking op until it is necessary.